### ISSUE TITLE
[=BUG=] TCP connection does not end.

### ISSUE THREAD

**Issue by cjh980402** [play-dl/play-dl#28]
#### Describe the bug

 - Node.js version: 16.6.2
 - OS & version: ubuntu 20.04

This is the result of 'lsof -i -a' command that song is normally finished.
```
COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node    10274 ubuntu   20u  IPv4 161808      0t0  TCP ubuntu-vm:40644->162.159.138.232:https (ESTABLISHED)
node    10274 ubuntu   21u  IPv4 161828      0t0  TCP ubuntu-vm:39840->162.159.135.232:https (ESTABLISHED)
node    10274 ubuntu   23u  IPv4 161814      0t0  TCP ubuntu-vm:58812->162.159.130.234:https (ESTABLISHED)
node    10274 ubuntu   24u  IPv4 162167      0t0  TCP ubuntu-vm:37216->162.159.130.235:https (ESTABLISHED)
node    10274 ubuntu   25u  IPv4 162168      0t0  UDP *:52051 
```
This is the result of 'lsof -i -a' command that song is forcely finished when song played about 20 seconds.
```
COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node    10274 ubuntu   20u  IPv4 161808      0t0  TCP ubuntu-vm:40644->162.159.138.232:https (ESTABLISHED)
node    10274 ubuntu   21u  IPv4 161828      0t0  TCP ubuntu-vm:39840->162.159.135.232:https (ESTABLISHED)
node    10274 ubuntu   23u  IPv4 161814      0t0  TCP ubuntu-vm:58812->162.159.130.234:https (ESTABLISHED)
node    10274 ubuntu   26u  IPv4 161790      0t0  TCP ubuntu-vm:56754->kix03s05-in-f11.1e100.net:https (ESTABLISHED)
```
This is the result of 'lsof -i -a' command that after 5minutes of second result.
```
COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node    10274 ubuntu   20u  IPv4 161808      0t0  TCP ubuntu-vm:40644->162.159.138.232:https (ESTABLISHED)
node    10274 ubuntu   21u  IPv4 161828      0t0  TCP ubuntu-vm:39840->162.159.135.232:https (ESTABLISHED)
node    10274 ubuntu   23u  IPv4 161814      0t0  TCP ubuntu-vm:58812->162.159.130.234:https (ESTABLISHED)
node    10274 ubuntu   26u  IPv4 161790      0t0  TCP ubuntu-vm:56754->kix03s05-in-f11.1e100.net:https (CLOSE_WAIT)
```
I found the bug that youtube TCP connection(kix03s05-in-f11.1e100.net:https) does not end when song is forcely finished.
#### Code to reproduce

```js

```

#### Checklist

- [o] I have read the documentation.
- [o] I have tried my code with the latest version of Node.js and Play-dl.

**Comment by killer069** [comment]
This doesn't close immediately but after few seconds of force closing.

Maybe try doing this after 10 seconds. Since I work with loops

**Comment by cjh980402** [comment]
If normal, the third result is not CLOSE_WAIT like first result(no youtube TCP). But TCP connection is until alive the process is terminated with CLOSE_WAIT status.

**Comment by killer069** [comment]
@cjh980402 
Can you try to clone developer branch and compile it and move dist folder to node_modules/play-dl/ ??

And test if it still exists or not ??

**Comment by cjh980402** [comment]
```
COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node    12701 ubuntu   20u  IPv4 221170      0t0  TCP ubuntu-vm:57658->162.159.137.232:https (ESTABLISHED)
node    12701 ubuntu   23u  IPv4 221174      0t0  TCP ubuntu-vm:54050->162.159.135.234:https (ESTABLISHED)
node    12701 ubuntu   24u  IPv4 225284      0t0  TCP ubuntu-vm:40512->162.159.136.232:https (ESTABLISHED)
node    12701 ubuntu   26u  IPv4 225335      0t0  TCP ubuntu-vm:46674->kix03s10-in-f7.1e100.net:https (ESTABLISHED)
```
It still exists in developer branch.

**Comment by killer069** [comment]
> ```
> COMMAND   PID   USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
> node    12701 ubuntu   20u  IPv4 221170      0t0  TCP ubuntu-vm:57658->162.159.137.232:https (ESTABLISHED)
> node    12701 ubuntu   23u  IPv4 221174      0t0  TCP ubuntu-vm:54050->162.159.135.234:https (ESTABLISHED)
> node    12701 ubuntu   24u  IPv4 225284      0t0  TCP ubuntu-vm:40512->162.159.136.232:https (ESTABLISHED)
> node    12701 ubuntu   26u  IPv4 225335      0t0  TCP ubuntu-vm:46674->kix03s10-in-f7.1e100.net:https (ESTABLISHED)
> ```
> 
> It still exists in developer branch.

Don't test it immediately, do it after 1 minute.

**Comment by cjh980402** [comment]
It's been over five minutes when I commented.

**Comment by killer069** [comment]
> It's been over five minutes when I commented.

So you mean you tested this after 5 minutes of force finishing a song ??

**Comment by cjh980402** [comment]
That's right. Also it still exists now.

**Comment by killer069** [comment]
> That's right. Also it still exists now.

I need some more testing before fixing this.

Try to use only video_info and get info from the url.

Then try to see if the issue comes or not.

Then try to create stream from url and check again.

I want to see if it is coming from video_info or the stream

**Comment by cjh980402** [comment]
```js
const youtube = require('play-dl');
(await youtube.video_info(url));
```
When I try above code, tcp connection is disappear immediatly.

```js
const youtube = require('play-dl');
(await youtube.stream(url, true));
```
When I try above code, tcp connection is disappear after some minutes.

```js
const { stream, type } = await youtube.stream(url, true);
createAudioResource(stream, {
    inputType: type,
    inlineVolume: true
});
```
However, when I play the song using AudioResource in above code, tcp connection is still exists after I stop the song.

**Comment by killer069** [comment]
> ```js
> const youtube = require('play-dl');
> await youtube.video_info(url);
> ```
> 
> When I try above code, tcp connection is disappear immediatly.
> 
> ```js
> const youtube = require('play-dl');
> await youtube.stream(url, true);
> ```
> 
> When I try above code, tcp connection is disappear after some minutes.
> 
> ```js
> const { setTimeout } = require('timers/promises');
> const youtube = require('play-dl');
> 
> const stream = await youtube.stream(url, true);
> const resource = createAudioResource(stream.stream, {
>     inputType: stream.type,
>     inlineVolume: true
> });
> player.play(resource);
> await setTimeout(10000);
> player.stop(true);
> ```
> 
> However, when I play the song using AudioResource in above code, tcp connection is still exists after I stop the song.

The video that you played is longer than 5 mins or not ??

**Comment by cjh980402** [comment]
shorter than 5mins

**Comment by killer069** [comment]
> Shorter than 5mins.
> If I play song for about a minute(1~1.5minute), the TCP connection disappears before the end of the song, but if I force the song to end before this state, the TCP connection remains.

The disappearance of TCP connection after a minute states that all data from the url has been passed to PassThrough. Pushed one more commit, check if now error occurs or not.

**Comment by cjh980402** [comment]
TCP connection error is almost same like before. However, when I use `await stream(<url>, true)` error check option, too many times are required than before.

**Comment by killer069** [comment]
> TCP connection error is almost same like before. However, when I use `await stream(<url>, true)` error check option, too many times are required than before.

Too many times are required, what does that mean ??

**Comment by cjh980402** [comment]
It takes almost a 60sec~90sec to get a YouTube stream, when I use `await stream(<url>, true)` error check option.

**Comment by cjh980402** [comment]
```js
const youtube = require('play-dl');
const fetch = require('node-fetch');
const got = require('got');

const start1 = Date.now();
let info = (await youtube.video_info('https://www.youtube.com/watch?v=YfrW2D766YA'));
let response = await fetch(info.format[info.format.length - 1].url);
const status1 = response.status;
const finish1 = Date.now();

const start2 = Date.now();
info = (await youtube.video_info('https://www.youtube.com/watch?v=YfrW2D766YA'));
response = await got(info.format[info.format.length - 1].url)
const status2 = response.statusCode;
const finish2 = Date.now();

const result = `node-fetch(${status1}): ${finish1-start1}, got(${status2}): ${finish2-start2}`;
```
I test this code, then I get the result `node-fetch(200): 1347, got(200): 22387`.

**Comment by killer069** [comment]
> ```js
> const youtube = require('play-dl');
> const fetch = require('node-fetch');
> const got = require('got');
> 
> const info = (await youtube.video_info('https://www.youtube.com/watch?v=YfrW2D766YA'));
> 
> const start1 = Date.now();
> const response1 = await fetch(info.format[info.format.length - 1].url);
> const status1 = response1.status;
> const finish1 = Date.now();
> 
> const start2 = Date.now();
> const response2 = await got(info.format[info.format.length - 1].url)
> const status2 = response2.statusCode;
> const finish2 = Date.now();
> 
> const result = `node-fetch(${status1}): ${finish1-start1}, got(${status2}): ${finish2-start2}`;
> ```
> 
> I test this code, then I get the result `node-fetch(200): 298, got(200): 21407`.

I forgot to add range headers in that.

Pushed one more commit, tell me if search is fixed or not.

**Comment by killer069** [comment]
I fixed the socket Error too. But Now you will get ECONNRESET Error ( if you force close a stream ).

**Comment by cjh980402** [comment]
> I forgot to add range headers in that.
> 
> Pushed one more commit, tell me if search is fixed or not.
This commit fixed the 'takes almost a 20sec~40sec to get a YouTube stream' issue. Thanks.

> I fixed the socket Error too. But Now you will get ECONNRESET Error ( if you force close a stream ).
Then, how can I handled ECONNRESET Error?

**Comment by cjh980402** [comment]
I tested more some code and case, I erase `this.socket?.destroy()` in cleanup and just add `stream.once('error', () => { stream.destroy(); });` in Stream.normal_start(), Stream.loop_start(), and Stream.loop() for normal song case, socket Error is fixed too.

**Comment by cjh980402** [comment]
I tested more some code and case. When I add `stream.once('error', () => {});` in Stream.normal_start(), Stream.loop_start(), and Stream.loop() for normal song case and erase code about socket code, I can ignore ECONNRESET Error and socket Error is fixed too.

**Comment by killer069** [comment]
> Also I think `stream.once('error', () => { stream.destroy(); });` is better.

Ok, I was testing some things. I got a very good solution.

I pushed the changes, now test if any error occurs or not. [ I want to publish 0.6.9 ]

**Comment by killer069** [comment]
I have made 404 error check to be enabled by default, no need to pass true now.

Check [this](https://github.com/play-dl/play-dl/tree/developer#streamurl--string)

**Comment by killer069** [comment]
Released 0.6.9

Fixed all errors

**Comment by cjh980402** [comment]
Thanks! All errors is fixed in 0.6.9!
